packages/pci-sdk-iframe/src/services/twoFactorAuth.service.ts

```diff
@@ -0,0 +1,88 @@
+import apiClient, { IVerify2FACodeResponse } from 'apiClient';
+import IApplicationState from 'types/IApplicationState';
+import ICardData from 'types/ICardData';
+import IStateFn from 'types/IStateFn';
+import { IStaticState } from 'types/IStaticState';
+import cardDataService from './cardData.service';
+
+interface IVerify2FACodeArgs {
+	secret: string;
+	dispatch: IStateFn;
+	state: IApplicationState;
+	staticState: IStaticState;
+}
+
+function verify2FACode(args: IVerify2FACodeArgs) {
+	return apiClient
+		.verify2FACode(args.secret, args.state.verificationId)
+		.then((res) => _onVerificationReceived({ res, ...args }))
+		.catch(() => args.dispatch({ isLoading: false, message: 'Unexpected error' }));
+}
+
+interface IOnVerificationReceivedArgs {
+	res: IVerify2FACodeResponse;
+	dispatch: IStateFn;
+	state: IApplicationState;
+	staticState: IStaticState;
+}
+
+// Callback executed when we receive a verification response from the server.
+function _onVerificationReceived({ res, dispatch, state, staticState }: IOnVerificationReceivedArgs) {
+	switch (res.status) {
+		// 2FA token is valid. We are good to get card data using the validated secret
+		case 'passed':
+			if (state.nextStep === 'SET_PIN') {
+				return dispatch({ uiStatus: 'SET_PIN_FORM' });
+			}
+
+			if (state.nextStep === 'VIEW_CARD_DATA') {
+				return cardDataService
```

Comment:
I don't like that this service has a dependency on another service. I thought to pass this callback as an arg to verify2FACode, but I don’t like declaring this function in useApp either 🤷🏻